### PR TITLE
Add a parser for 2bit files to Bio.SeqIO

### PR DESCRIPTION
**Pull request by mdehoon**
- [X] I hereby agree to dual licence this and any previous contributions under both
the _Biopython License Agreement_ **AND** the _BSD 3-Clause License_.

- [X] I have read the ``CONTRIBUTING.rst`` file, have run ``pre-commit`` locally,
and understand that AppVeyor and TravisCI will be used to confirm the Biopython unit
tests and style checks pass with these changes.

- [X] I have added my name to the alphabetical contributors listings in the files
``NEWS.rst`` and ``CONTRIB.rst`` as part of this pull request, am listed
already, or do not wish to be listed. (*This acknowledgement is optional.*)

This PR adds a parser for twoBit (.2bit) DNA sequence files to `Bio.SeqIO`.
This format is used by UCSC to distribute large genome data (including the human genome) as a single file for one genome.
TwoBit files contain an index, allowing fast random access to the genome sequence.
The parser takes advantage of this index as follows:
```python
>>> from Bio import SeqIO
>>> records = SeqIO.parse("hg38.2bit", "twobit") # this takes about 8 seconds to read the index
>>> records.keys()
('chr1', 'chr10', 'chr11', 'chr11_KI270721v1_random', 'chr12', 'chr13', 'chr14', 'chr14_GL000009v2_random', ...
>>> record = records['chr12']
>>> record.id
'chr12'
>>> seq = record.seq
>>> seq[44000000:44000100]  # this is now instantaneous
Seq('ATCATCAATTTTGCCCAATGatcagctgactaattgatgaagtccttcttcagt...ttc')
```

Otherwise, the parser works the same as the other parsers in `Bio.SeqIO`.

### CODE REVIEW

**Comment by peterjc** [comment]
Rather than adding a ``.keys()`` method, or in additiona, can you do the ``Bio.SeqIO.index()`` integration? SFF is similar in that by default there is a built in index.

**Comment by mdehoon** [comment]
@peterjc  Sorry I don't think I quite understand what `SeqIO.index` would add beyond what `SeqIO.parse` is already offering.

**Comment by mdehoon** [comment]
@peterjc  I had a look at `SeqIO.index` as I was not familiar with this function. If I understand correctly, it works a bit differently than the twoBit parser and stores the location in the file of each record in an index. When a user tries to access this record through the dictionary interface, we jump to the file position stored in the index, parse the data, and creates the `Seq` and `SeqRecord` objects. Correct?

The twoBit parser works differently: All `Seq` objects are created and stored when the parser reads the index; the data argument passed to the `Seq` initializer is not a `bytes` object but a bytes-like object. When the user asks for a specific sequence region, e.g. `record.seq[100:200]`, the bytes-like object reads that specific sequence region from the file and returns it.

The reason it was designed this way is that twoBit files typically contain few sequences (e.g. equal to the number of chromosomes), but each sequence is very large, so you don't want to read the whole sequence just to get a short segment. Sff files on the other hand contain short sequences, but there are lots (many millions) of them, so you don't want to create all sequences in advance.

As a second point: It was straightforward to add a `keys` and `__getitem__` method to the twoBit iterator, and it may be possible to use the same strategy to `SffIterator` et al. Then you won't need a separate `index` function. Or you could let the iterator do the indexing automatically the first time a user calls `keys` or `__getitem__` on the `SffIterator`. But anyway, that is beyond this PR.

**Comment by peterjc** [comment]
It is possible for the twoBit parser to avoid loading everything into RAM, or is this a big issue on the typical file sizes and usage?

If you do just load it all into RAM, then ``Bio.SeqIO.index`` does not add anything.

**Comment by mdehoon** [comment]
@peterjc  Everything can easily be loaded into RAM. For example for the human genome it's only 455 sequences.

**Comment by peterjc** [comment]
Easy point first: There is nothing in the parser comments pointing at the file format, or any comments about the file format structure.

General question: Was it really necessary to write this in C? I would have thought Python sufficient and easier for long term maintenance.

Technical question: I see you are using ``lseek``, but because you are using C this presumably bypasses the forward only testing. This presumably prevents parsing stdin? Is it not possible to parse the file in a single linear read through the file - that would be in keeping with how the SeqIO parsers work in general?

**Comment by peterjc** [comment]
Is http://genome.ucsc.edu/FAQ/FAQformat.html#format7 the official format definition? Quoting:

> # .2bit format
> A .2bit file stores multiple DNA sequences (up to 4 Gb total) in a compact randomly-accessible format. The file contains masking information as well as the DNA itself.
> 
> The file begins with a 16-byte header containing the following fields:
> 
> * signature - the number 0x1A412743 in the architecture of the machine that created the file
> * version - zero for now. Readers should abort if they see a version number higher than 0
> * sequenceCount - the number of sequences in the file
> * reserved - always zero for now
>
> All fields are 32 bits unless noted. If the signature value is not as given, the reader program should byte-swap the signature and check if the swapped version matches. If so, all multiple-byte entities in the file will have to be byte-swapped. This enables these binary files to be used unchanged on different architectures.
> 
> The header is followed by a file index, which contains one entry for each sequence. Each index entry contains three fields:
> 
> * nameSize - a byte containing the length of the name field
> * name - the sequence name itself (in ASCII-compatible byte string), of variable length depending on nameSize
> * offset - the 32-bit offset of the sequence data relative to the start of the file, not aligned to any 4-byte padding boundary
>
> The index is followed by the sequence records, which contain nine fields:
> 
> * dnaSize - number of bases of DNA in the sequence
> * nBlockCount - the number of blocks of Ns in the file (representing unknown sequence)
> * nBlockStarts - an array of length nBlockCount of 32 bit integers indicating the (0-based) starting position of a block of Ns
> * nBlockSizes - an array of length nBlockCount of 32 bit integers indicating the length of a block of Ns
> * maskBlockCount - the number of masked (lower-case) blocks
> * maskBlockStarts - an array of length maskBlockCount of 32 bit integers indicating the (0-based) starting position of a masked block
> * maskBlockSizes - an array of length maskBlockCount of 32 bit integers indicating the length of a masked block
> * reserved - always zero for now
> * packedDna - the DNA packed to two bits per base, represented as so: T - 00, C - 01, A - 10, G - 11. The first base is in the most significant 2-bit byte; the last base is in the least significant 2 bits. For example, the sequence TCAG is represented as 00011011.
>
> For a complete definition of all fields in the twoBit format, see this description in the source code.

**Comment by peterjc** [comment]
Reading that, I would parse the header into memory (including all the names), and then loop over the sequence records without seeking. Index support could work very much like the SFF work.

**Comment by mdehoon** [comment]
> Is http://genome.ucsc.edu/FAQ/FAQformat.html#format7 the official format definition?

Yes.

**Comment by mdehoon** [comment]
> General question: Was it really necessary to write this in C? I would have thought Python sufficient and easier for long term maintenance.

Python would be slow. Converting 2bit to sequences means converting pairs of 2 bits to nucleotides; it's not something I would like to do in Python.

**Comment by mdehoon** [comment]
> Reading that, I would parse the header into memory (including all the names), and then loop over the sequence records without seeking. Index support could work very much like the SFF work.

See my previous description at https://github.com/biopython/biopython/pull/3388#issuecomment-731600334 . The biggest issue is not to find the start of the sequence record (like for SFF), but how to only read the part of the sequence the user is interested in. Otherwise, we'd have to read in the full chromosome each time, which will require much more memory.

**Comment by mdehoon** [comment]
> Technical question: I see you are using lseek, but because you are using C this presumably bypasses the forward only testing. This presumably prevents parsing stdin?

It is possible to parse stdin:

```
$ cat test.py 
import sys
from Bio import SeqIO

records = SeqIO.parse(sys.stdin.buffer, "twobit")
for record in records:
    print(record.id, len(record.seq), record.seq[:10])

$ python test.py < hg38.2bit   
chr1 248956422 NNNNNNNNNN
chr10 133797422 NNNNNNNNNN
chr11 135086622 NNNNNNNNNN
chr11_KI270721v1_random 100316 ctgcggtcag
chr12 133275309 NNNNNNNNNN
chr13 114364328 NNNNNNNNNN
chr14 107043718 NNNNNNNNNN
chr14_GL000009v2_random 201709 GAATTCTGTG
chr14_GL000225v1_random 211173 GAattctcac
chr14_KI270722v1_random 194050 GATCCATATA
chr14_GL000194v1_random 191469 GATCCCTGCC
...
```
I am not sure though why you would want to do that.

> Is it not possible to parse the file in a single linear read through the file - that would be in keeping with how the SeqIO parsers work in general?

Yes you can, e.g.
```python
from Bio import SeqIO

handle = open("Downloads/hg38.2bit", "rb")
records = SeqIO.parse(handle, "twobit")
d = {}
for record in records:
    d[record.id] = record.seq[:]
```

But you'd have to hold sequences of all chromosomes in memory, i.e. around 3Gb for human.

**Comment by mdehoon** [comment]
Some timings:

```python
handle = open("Downloads/hg38.2bit")
records = SeqIO.parse(handle, "twobit")
for record in records:
    s = record.seq[:]
    print(record.id, len(s))
handle.close()
```

9.4 seconds

```python
handle = open("Downloads/hg38.fa")
records = SeqIO.parse(handle, "fasta")
for record in records:
    s = record.seq[:]
    print(record.id, len(s))
handle.close()
```

52.4 seconds

**Comment by mdehoon** [comment]
Some more timings:
```python
from Bio import SeqIO

handle = open("Downloads/hg38.2bit", "rb")

records = SeqIO.parse(handle, "twobit")
# 7.1 seconds

print(records['chr1'].seq[200000000:200000100])
# 0.00037 seconds

print(records['chr1'].seq[200000000:200000100])
# 0.00032 seconds

print(records['chrX'].seq[100000000:100000100])
# 0.00018 seconds
handle.close()
```

Same thing by indexing a Fasta file:
```python
records = SeqIO.index("Downloads/hg38.fa", "fasta")
# 135.2 seconds

print(records['chr1'].seq[200000000:200000100])
# 8.59 seconds

print(records['chr1'].seq[200000000:200000100])
# 8.62 seconds

print(records['chrX'].seq[100000000:100000100])
# 5.50 seconds
```

File: Bio/Seq.py [hunk 1]
```diff
@@ -134,7 +149,10 @@ def __str__(self):
                 as_string = str(seq_obj)
 
         """
-        return self._data.decode("ASCII")
+        data = self._data
+        if not isinstance(data, (bytes, bytearray)):
```
Comment (peterjc) [inline Bio/Seq.py:153]: What else could it be?

**Comment by peterjc** [comment]
Can you add http://genome.ucsc.edu/FAQ/FAQformat.html#format7 to the docstring then?

Good to know stdin would work, although I don't immediately have a use case, it is often helpful when using SeqIO at the command line.

I would have thought that the same mapping of bytes to four bases at a time trick would have worked as well in Python as in C. Did you go straight to C for this parser? If not, writing a new version in Python does seem unnecessary. I suppose since the format has been stable for years, any maintenance costs to the C code are more going to be able things like Python and CPU changes.

Adding ``Bio.SeqIO.index`` support should be very easy, and I would prefer that over the 2bit specific dictionary like behaviour you are proposing.

File: Bio/Seq.py (hunk 1 above)
Comment (mdehoon) [reply Bio/Seq.py:153]: Any bytes-like object, meaning an object that returns `bytes` when calling `__getitem__` on it.
Currently in Biopython, only `TwoBitSequence` objects do that. This allows them to return subsequences without having to read the whole chromosome. I can imagine other uses for this in Biopython, for example for `UnknownSeq` and `DBSeq` in `BioSQL`.

Comment (peterjc) [reply Bio/Seq.py:153]: Interesting, and I start to see what you mean for ``UnknownSeq`` and ``DBSeq`` now.

I would suggest added a comment here, ``# e.g. the TwoBitSequence objects which can act like a view onto a sub-sequence``.

Having the 2bit C code start with some explanatory human readable text would with understanding the design. I had missed the significance of ``TwoBitSequence`` completely.

File: Bio/Seq.py [hunk 2]
```diff
@@ -95,26 +95,41 @@ def __init__(self, data):
             self._data = bytes(data)
         elif isinstance(data, str):
             self._data = bytes(data, encoding="ASCII")
+        elif self._check_bytes_like(data):
+            self._data = data
```
Comment (peterjc) [inline Bio/Seq.py:99]: I would suggest added a comment here, something like:

``# e.g. the TwoBitSequence objects which can act like a view onto a sub-sequence``

Comment (mdehoon) [reply Bio/Seq.py:99]: We may want to define a abstract base class (as a separate PR) to specify what exactly we mean by a bytes-like object here. Python itself has a similar concept (anything that implements the buffer interface for bytes).

Comment (peterjc) [reply Bio/Seq.py:99]: Sounds sensible.

**Comment by mdehoon** [comment]
> I would have thought that the same mapping of bytes to four bases at a time trick would have worked as well in Python as in C. Did you go straight to C for this parser? If not, writing a new version in Python does seem unnecessary. I suppose since the format has been stable for years, any maintenance costs to the C code are more going to be able things like Python and CPU changes.

I went straight to C for this parser. The bit-twiddling that is needed in this parser (not just for the sequence, also for the index) is easier in C. I don't think the format will change much, and there is not much Python-C glue code needed.

**Comment by mdehoon** [comment]
> Adding Bio.SeqIO.index support should be very easy, and I would prefer that over the 2bit specific dictionary like behaviour you are proposing.

The `Bio.SeqIO.index` function made sense when `Bio.SeqIO.parse` was a generator function. But now that `Bio.SeqIO.parse` returns an iterator, it's trivial to add `__getitem__`, `__len__`, and `keys` to the iterator and give it dictionary-like capabilities. You would just need to trigger the indexing to happen the first time any of these three functions is called. Then we can get rid of `Bio.SeqIO.index`.

**Comment by peterjc** [comment]
Better to revert 01d1af8b6776ba14c6460c33e095a30bd2d5ce03 and correctly format the text as RST. Try https://livesphinx.herokuapp.com/ for this?

Otherwise the documentation is invisible as a comment, and will not appear as a docstring on our website etc.

**Comment by peterjc** [comment]
We'd need to be more explicit about the handle ownership assuming if we go down the route of adding dictionary-like behaviour to the parsers directly. Right now that's something else your end user documentation could be clear about?

**Comment by mdehoon** [comment]
> We'd need to be more explicit about the handle ownership assuming if we go down the route of adding dictionary-like behaviour to the parsers directly.

Not sure what you mean by this. The handle should not be closed, but that was always the case.

**Comment by peterjc** [comment]
One of the design choices with Bio.SeqIO.index was not to take a file handle, only file names - precisely because it has to own the handle and will jump and seek (and close and reopen handles too when multiple files are indexed).

On the other hand, straight forward looping over a file like this has until now worked, regardless of the format:

```python
favourite_records = []
handle = open(filename)
for record in SeqIO.parse(handle, format):
    if len(record) < 100:
        favourite_records.append(record)
handle.close()  # this SHOULD be closed now
# Can still use anything in favourite_records
```

or,

```python
favourite_records = []
with open(filename) as handle:
    for record in SeqIO.parse(handle, format):
        if len(record) < 100:
            favourite_records.append(record)
# Can still use anything in favourite_records
```

That will break with the current implementation of your 2bit parser, won't it? This is a downside to the lazy loading. This needs to be clearer.

**Comment by mdehoon** [comment]
@peterjc  Thank you for the explanation. I have added some description to the documentation.

> One of the design choices with Bio.SeqIO.index was not to take a file handle, only file names - precisely because it has to own the handle and will jump and seek (and close and reopen handles too when multiple files are indexed).

I am not sure what is a better approach here. The alternative is to allow file handles only, and not file names. With a file name, you need an additional `close` function to close the file handle, as in the example:
```python
>>> from Bio import SeqIO
>>> record_dict = SeqIO.index("Fasta/f002", "fasta")
>>> len(record_dict)
3
>>> print(len(record_dict["gi|1348917|gb|G26685|G26685"]))
413
>>> record_dict.close()
>>> print(len(record_dict["gi|1348917|gb|G26685|G26685"]))
Traceback (most recent call last):
...    handle.seek(offset)
ValueError: seek of closed file
```
With a file handle, you would get the following:
```python
>>> from Bio import SeqIO
>>> handle = open("Fasta/f002")
>>> record_dict = SeqIO.index(handle, "fasta")
>>> len(record_dict)
3
>>> print(len(record_dict["gi|1348917|gb|G26685|G26685"]))
413
>>> handle.close()
>>> print(len(record_dict["gi|1348917|gb|G26685|G26685"]))
Traceback (most recent call last):
...    handle.seek(offset)
ValueError: seek of closed file
```
which doesn't require a new `close` function. As long as the error message is clear ("seek of closed file" is not optimal), both are fine.

The twoBit parser currently allows both file handles and file names. In the latter case, the file is closed at some point after the records go out of scope (immediately for C-python, but delayed for PyPy depending on the garbage collector). This is not ideal, but neither is adding a `close` function to `TwoBitIterator`.

This is what it looks like with a handle:
```python
>>> from Bio import SeqIO
>>> handle = open("sequence.bigendian.2bit", "rb")
>>> records = SeqIO.parse(handle, "twobit")
>>> handle.close()
>>> records['seq222'].seq
Traceback (most recent call last):
...
ValueError: cannot retrieve sequence: file is closed
```

Another option is to allow both file names and file handles, and explain in the documentation that file handles should be used if the user wants to have control over when exactly the file should be closed.

**Comment by peterjc** [comment]
It is nice to see the lazy loading aspect of this, given our GSoC work with @eparker05 on #356. If we restrict ourselves to lazy loading the sequence for now, then FASTA when an ``*.fai`` index is present is another strong candidate.

If some of the SeqIO parsers are going to become lazy loading, then adding a close method seems reasonable for explicitly closing the handles.

People will run into the OS max open handle limit, and there has to be a way to deal with this. We might consider a SeqIO wide handle pool, similar to the per-index pool used when indexing multiple files (see ``max_open`` in ``Bio/File.py``). The simpler route is to provide a mechanism for the user to control this by explicitly closing handles - for which adding a close method seems reasonable. The other reason I added the ``.close()`` method for my indexing code was to allow the handles to be closed explicitly (in our tests in particular), in order to silence warnings from the PyPy tests about potential resource leaks.

The mode complication is another good reason to encourage the SeqIO functions be used primarily with filenames - it lets us pick text mode or binary mode automatically, e.g. for a faster FASTA or FASTQ parser, even if we have to keep the text mode support alive during a deprecation period. i.e. Demote use of a handle to more advanced usage? Right now use cases are compressed files (e.g. using ``gzip.open``) and hybrid files (e.g. embedding a plain text format within HTML or markdown).

**Comment by mdehoon** [comment]
Updated timings:

```python
from Bio import SeqIO

handle = open("Downloads/hg38.2bit", "rb")

records = SeqIO.parse(handle, "twobit")
# 0.11 seconds

print(records['chr1'].seq[200000000:200000100])
# 0.0015 seconds

print(records['chr1'].seq[200000000:200000100])
# 0.00054 seconds

print(records['chrX'].seq[100000000:100000100])
# 0.00040 seconds
handle.close()
```
Reading the index is much faster (as numpy will read a complete array in one go), but retrieving the sequence is about 2-3 times slower than the C version.

**Comment by mdehoon** [comment]
@peterjc 

> The simpler route is to provide a mechanism for the user to control this by explicitly closing handles - for which adding a close method seems reasonable. The other reason I added the .close() method for my indexing code was to allow the handles to be closed explicitly (in our tests in particular), in order to silence warnings from the PyPy tests about potential resource leaks.

The return value from `SeqIO.parse` has an attribute `.stream` which is the file handle. So to close the handle, you can do
```python
>>> from Bio import SeqIO
>>> records = SeqIO.parse("sequence.fa", "fasta")
>>> records.stream.close()
# confirm it's really closed:
>>> for record in records:
...     pass
... 
Traceback (most recent call last):
...
ValueError: I/O operation on closed file.
>>> 

```
without having to add a new `.close` method.

**Comment by peterjc** [comment]
You probably used ``.stream`` over ``.handle`` deliberately, didn't you? Even though ``.handle`` would be consistent with history usage elsewhere in Biopython.

**Comment by mdehoon** [comment]
@peterjc  Python uses "stream" in these cases. See for example the docstring for the `io` module.

**Comment by mdehoon** [comment]
This is still quite a bit slower than the C version:

```python
handle = open("Downloads/hg38.2bit", "rb")
records = SeqIO.parse(handle, "twobit")
for record in records:
    s = record.seq[:]
    print(record.id, len(s))
handle.close()
```
29.75 seconds

compared to 9.4 seconds for the C version, and 52.4 seconds for the equivalent on a Fasta file. Especially masking the sequence is slow if done in Python.

**Comment by mdehoon** [comment]
Using C code just for the 2bit to sequence conversion:

```python
handle = open("Downloads/hg38.2bit", "rb")
records = SeqIO.parse(handle, "twobit")
for record in records:
    s = record.seq[:]
    print(record.id, len(s))
handle.close()
```
2.7 seconds

**Comment by mdehoon** [comment]
@peterjc  Are you OK with this one, or do you have any remaining concerns?

File: Bio/SeqIO/TwoBitIO.py [hunk 3]
```diff
@@ -0,0 +1,229 @@
...
+
+
+class TwoBitSequenceData:
+    """Store the information needed to retrieve sequence data from a .2bit file.
+
+    Objects of this class store the file position at which the sequence data
+    start, the sequence length, and the start and end position of unknown (N)
+    and masked (lowercase) letters in the sequence.
+
+    Only two methods are provided: __len__ and __getitem__. The former will
```
Comment (peterjc) [inline Bio/SeqIO/TwoBitIO.py:94]: Is there value in defining ``__bytes__`` too?

Can you not define ``__repr__`` and perhaps even ``__str__`` as well?

File: Bio/SeqIO/TwoBitIO.py [hunk 4]
```diff
@@ -0,0 +1,229 @@
...
+        byteEnd = (end + 3) // 4
+        byteSize = byteEnd - byteStart
+        stream = self.stream
+        try:
+            stream.seek(self.offset + byteStart)
+        except ValueError as exception:
+            if str(exception) == "seek of closed file":
+                raise ValueError("cannot retrieve sequence: file is closed") from None
+            raise
+        data = numpy.fromfile(stream, dtype="uint8", count=byteSize)
```
Comment (peterjc) [inline Bio/SeqIO/TwoBitIO.py:127]: Why bother with numpy at all? Why not pass the raw bytes to the C code?

**Comment by peterjc** [comment]
Design question: Why use numpy when as far as I can see, Python's struct module (as used in the SFF parser) would work just as well

**Comment by peterjc** [comment]
Can you add ``Bio.SeqIO.index`` support as well? Or do you want me to add it afterwards?

File: Bio/SeqIO/TwoBitIO.py (hunk 3 above)
Comment (mdehoon) [reply Bio/SeqIO/TwoBitIO.py:94]: This class is not meant to be used by the end user, but only as the `data` argument when creating `Seq` object by `TwoBitIterator`. Therefore we want to create the absolute minimal class here. The `Seq` object will store the `data` argument as the `_data` attribute, but this attribute is not meant to be accessed by the user directly, only via the `Seq` object methods.

**Comment by mdehoon** [comment]
> Design question: Why use numpy when as far as I can see, Python's struct module (as used in the SFF parser) would work just as well

That may work. But can you efficiently read arrays with the struct module? For example maskBlockStarts is an array consisting 10,000's of integers.

**Comment by mdehoon** [comment]
> Can you add Bio.SeqIO.index support as well? Or do you want me to add it afterwards?

I am not convinced that `Bio.SeqIO.index` is needed for any format; you could just add `keys` and `__getitem__` to the iterator returned by `SeqIO.parse`, and have it run the indexing code the first time either of those two methods are used.

**Comment by peterjc** [comment]
Using struct to get lots of integers at once works, e.g.
https://github.com/biopython/biopython/blob/biopython-178/Bio/SeqIO/SffIO.py#L697 - not sure if it would be any faster.

--

For most file formats there is a significant cost to indexing (2bit, SFF and FAI are exceptions). Your proposal would work, and doing this to all the existing indexed file formats would make ``Bio.SeqIO.index`` redundant. The multiple-file use case for ``Bio.SeqIO.index_db`` remains though.

If we are going to add indexing and lazy-loading via ``Bio.SeqIO.parse`` (and ``Bio.SeqIO.read`` where this does still make sense for large chromosomes), I think it should be an explicit optional argument, and that could require a filename not a stream/handle. The default (at least for existing parsers) should be no lazy-loading for backwards compatibility. For 2bit, it could default to True (and require it to be True).

For more lazy loading support, SFF would be a logical next choice, in part because it has a built in index, or FASTA with an FAI file. I'd be tempted to do GenBank/EMBL after that.

**Comment by mdehoon** [comment]
> Using struct to get lots of integers at once works, e.g.
https://github.com/biopython/biopython/blob/biopython-178/Bio/SeqIO/SffIO.py#L697 - not sure if it would be any faster.

Thanks. This is quite a bit slower: 1.1 seconds for reading the index for the human genome, compared to 0.11 seconds when using numpy. The reason for this is that `numpy.fromfile` directly reads the data into a numpy array. In contrast, with `struct` you first need to read the data into a `bytes` object, then use `struct.unpack` to convert that to a list, and then copy that list into a numpy array. (If you don't copy the data into either a numpy array or an `array.array`, then the C code has to pull each number separately out of the list and convert it into an integer, which is very slow and cumbersome to program. With a numpy array or an `array.array`, in the C code you directly get a pointer to the array of integers.)

**Comment by mdehoon** [comment]
Can we move ahead with this one?

File: Bio/SeqIO/TwoBitIO.py (hunk 3 above)
Comment (peterjc) [reply Bio/SeqIO/TwoBitIO.py:94]: Then give the class a leading underscore and add "(PRIVATE)" to the end of the class docstring?

**Comment by peterjc** [comment]
What do you think about changes to the ``Bio.SeqIO.parse`` and ``Bio.SeqIO.read`` API for controlling lazy loading? A per-format default is OK, although off by default is simpler to explain. Or are you indirectly advocating for this not being a user-setting but simply down to the file format? That would hinder adding lazy loading to existing parsers.

**Comment by mdehoon** [comment]
> What do you think about changes to the Bio.SeqIO.parse and Bio.SeqIO.read API for controlling lazy loading? A per-format default is OK, although off by default is simpler to explain. Or are you indirectly advocating for this not being a user-setting but simply down to the file format? That would hinder adding lazy loading to existing parsers.

For the 2bit format, it makes sense to always use lazy loading, as a user can always do `seq = seq[:]` to load the full sequence. But in general, I can imagine that `SeqIO.parse` takes some optional arguments that get passed to the parser and tell it how lazy it should be. The framework would be general but the specific arguments and their interpretation would depend on the file format.

**Comment by peterjc** [comment]
OK. Something in the NEWS file, and this should be OK to merge.

**Pull request merged**